alf/drivers/async_off_policy_driver.py

```diff
@@ -0,0 +1,220 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable
+from absl import logging
+from collections import namedtuple
+
+from alf.utils import common
+from alf.algorithms.off_policy_algorithm import OffPolicyAlgorithm
+from alf.algorithms.off_policy_algorithm import make_experience
+
+import gin.tf
+import tensorflow as tf
+
+from alf.drivers.off_policy_driver import OffPolicyDriver
+from alf.drivers.threads import TFQueues, ActorThread, EnvThread, LogThread
+from alf.experience_replayers.experience_replay import OnetimeExperienceReplayer
+
+LearningBatch = namedtuple(
+    "LearningBatch",
+    "time_step policy_step act_dist_param next_time_step env_id")
+
+
+def make_learning_batch(time_step, policy_step, act_dist_param, next_time_step,
+                        env_id):
+    return LearningBatch(
+        time_step=time_step,
+        policy_step=policy_step,
+        act_dist_param=act_dist_param,
+        next_time_step=next_time_step,
+        env_id=env_id)
+
+
+@gin.configurable
+class AsyncOffPolicyDriver(OffPolicyDriver):
+    """
+    A driver that enables async training. The definition of 'async' is that
+    prediction and learning are not intervened like below (synchronous):
+        pred -> learn -> pred -> learn -> ...
+    Instead they are decoupled:
+        pred -> pred -> pred -> ...
+                   |
+            (synchronize periodically)
+                   |
+        learn -> learn -> learn -> ...
+    And more importantly, a learner or predictor may only operate on a subset of
+    environments at a time.
+    """
+
+    def __init__(self,
+                 env_f: Callable,
+                 algorithm: OffPolicyAlgorithm,
+                 num_envs=1,
+                 num_actor_queues=1,
+                 unroll_length=8,
+                 learn_queue_cap=1,
+                 actor_queue_cap=1,
+                 observers=[],
+                 metrics=[],
+                 exp_replayer_class=OnetimeExperienceReplayer,
+                 debug_summaries=False,
+                 summarize_grads_and_vars=False,
+                 train_step_counter=None):
+        """
+        Args:
+            env_f (Callable): a function with 0 args that creates an environment
+            algorithm (OffPolicyAlgorithm):
+            num_envs (int): the number of environments to run asynchronously.
+                Note: each environment itself could be a tf_agent batched
+                environment. So the actual total number of environments is
+                `num_envs` * `env_f().batch_size`. However, `env_f().batch_size`
```

Comment:
Although you have the check of this condition in threads.py, it's better to have a check here since it will be easier for user to fix the parameter if the check fails. If they are in threads.py, user need to figure out where they are originally from.